sapicore/engine/synapse/__init__.py

```diff
@@ -0,0 +1,330 @@
+"""Synapses connect neurons or ensembles to each other."""
+import logging
+
+from collections import deque
+from typing import Callable
+
+import torch
+import numpy as np
+
+from torch import tensor, Tensor
+from torch.nn.init import xavier_uniform_
+
+from sapicore.engine.component import Component
+from sapicore.engine.neuron import Neuron
+
+__all__ = ("Synapse",)
+
+
+class Synapse(Component):
+    """Static synapse base class.
+
+    This class provides basic parameters and logic for connecting ensemble instances. Static synapses keep their
+    initial weights throughout the simulation unless manually adjusted. Synapse instances have 2D weight and
+    connectivity matrices, reflecting their role as layer-to-layer connectors.
+
+    Synapses own the following parameters, over and above those inherited from :class:`~engine.component.Component`:
+
+    Parameters
+    ----------
+    src_ensemble: Neuron
+        Reference to the presynaptic (sending) ensemble.
+
+    dst_ensemble: Neuron
+        Reference to postsynaptic ensemble (receiving) ensemble.
+
+    weight_max: float or Tensor
+        Positive weight value limit for any synapse element (applies throughout the simulation).
+
+    weight_min: float or Tensor
+        Negative weight value limit for any synapse element (applies throughout the simulation).
+
+    delay_ms: float or Tensor
+        Generic transmission delay, managed by the synapse object using queues.
+        Transmission delays are useful for some temporal coding schemes.
+
+    simple_delays: bool
+        Whether transmission delay values from one presynaptic element are the same for all postsynaptic targets.
+        Provides a speed-control tradeoff (one matrix multiplication vs. N vector multiplications).
+
+    weight_init_method: Callable
+        Weight initialization method from `torch.nn.init`.
+
+    connections: Tensor
+        2D binary integer mask matrix indicating which source unit -> destination unit connections are enabled.
+        Loggable attribute.
+
+    weights: Tensor
+        2D float matrix storing a weight for each pair of source and destination ensemble units.
+        Loggable attribute.
+
+    output: Tensor
+        1D tensor storing the result of W*x matrix multiplication in the current simulation step.
+        Loggable attribute.
+
+    """
+
+    _config_props_: tuple[str] = ("weight_max", "weight_min", "delay_ms", "simple_delays")
+    _loggable_props_: tuple[str] = ("weights", "connections", "output")
+
+    # declare loggable instance attributes registered as torch buffers.
+    connections: tensor
+    weights: tensor
+    output: tensor
+
+    def __init__(
+        self,
+        src_ensemble: [Neuron] = None,
+        dst_ensemble: [Neuron] = None,
+        weight_max: float = 1000.0,
+        weight_min: float = -1000.0,
+        delay_ms: float = 0.0,
+        simple_delays: bool = True,
+        weight_init_method: Callable = xavier_uniform_,
```

Comment:
But this can be left for later. As you have many places that accept callables - which would be a problem for repeatable experiments.

Reply:
Is there a reason to use the factory pattern here other than registering strings for logging/configuration? We are not abstracting away the instantiation of classes conditional on some value (as we would, e.g., with Integrators), just referencing existing torch or scipy functions.

I can imagine a user writing their own arbitrary weight initialization function or picking one of dozens of metric functions from scipy when computing distances. How would a factory pattern serve us there? In the most general case, if a user wanted to write a Pipeline with configurable Callables, they could dynamically import the Callable based on a string in a dict/YAML (following my network config file example).

What repeatability problems do you anticipate? If the Callable comes from an external dependency, providing the environment info with the code would keep things consistent. Also, the random seed can be fixed across libraries. Is there anything I'm missing?

Reply:
I guess the difference is how you think of `weight_init_method`. My implicit assumption was that it'd be part of the config, so that you have e.g. a config property named `weight_init` and based on that we select the `weight_init_method`. Rather than directly providing the method.

However, if you don't think that as part of the config, passing in the method is fine. But, if you did want to make it configurable then you'd need find a way to convert strings to functions. That's what you typically use a factory for. But I guess we don't need to worry about this right now.